Propose changes in the Function CRD (#15862)
kyma-project/kyma: components/function-controller/design/extend_function_spec_proposal.md

spec:
  serviceBindings:
  - source: my-secret
  env:
  - name: SERVICE_BINDING_ROOT 
    value: /foo
```

### Metadata for function pod

Define runtime labels and annotations on the root level or under a `metadata` field

```yaml
apiVersion: serverless.kyma-project.io/v1alpha2
kind: Function
metadata:
  name: my-function
  namespace: default
  labels:
    ...
  annotations:
    ...
spec:
  metadata: 
    labels: ...
    annotations: ...
```
OR
```yaml
apiVersion: serverless.kyma-project.io/v1alpha2
kind: Function
metadata:
  name: my-function
  namespace: default
  labels:
    ...
  annotations:
    ...
spec:
  labels: ...
  annotations: ...
```

## Samples: 

### Option 1

- simplified mounts serving just for service binding purpose
- extract `build` object for any build-time specific config
- runtime labels and annotations on the spec root level

```yaml
apiVersion: serverless.kyma-project.io/v1alpha2
kind: Function
metadata:
  name: my-function
  namespace: default
  labels:
    app.kubernetes.io/name: my-function
spec:
  runtime: nodejs16
  source:
    ...
  serviceBindings: 
  - source: my-secret
    mountPath: "/foo" # optional.. read from SERVICE_BINDING_ROOT ENV 

>   profile: S / M / L / XL / ... / Custom #optional

Reviewer: I would suggest thinking about the name of this field. I think that the `.spec.profile` name is too generic and does not introduce any context. What do you think about `resourcesProfile`?

- dif2: rename the `.spec.profile` to the `.spec.resourcesProfile` to make this field more intuitive
- dif3: this solution is simple but not that simple as the `Option 1`

```yaml
apiVersion: serverless.kyma-project.io/v1alpha2
kind: Function
metadata:
  name: my-function
  namespace: default
  labels:
    app.kubernetes.io/name: my-function
spec:
  runtime: nodejs16
  source:
    ...
  runtime:
    serviceBindings: 
    - source: my-secret
      mountPath: "/foo" # optional.. read from SERVICE_BINDING_ROOT ENV 

    resourcesProfile: S / M / L / XL / ... / Custom #optional
    resources: # optional... required if spec.profile==Custom
      limits: ... #if_custom
      requests: ... #if_custom

    labels: 
      app: my-app
    annotations: 
      fluentbit.io/parser: my-regex-parser
      istio-injection: enabled
  
    env:
    - name: SERVICE_BINDING_ROOT
      value: /service_bindings

  build:
    resourcesProfile: S / M / L / XL / ... / Custom
    resources: # optional... required if spec.build.profile==Custom
      limits: ... #if_custom
      requests: ... #if_custom

```

> ### Option 3

Reviewer 2: I reserved it :)

> ### Option 4

Reviewer 3: reserved

# Extending CRD spec for Serverless v1alpha2

## Summary

The current Serverless API allows for limited configuration of the generated Function's deployment. 
Currently users can only use ENVs to pass 3rd party service credentials but it doesn't allow for volume mounted secrets (which becomes the industry standard for [service bindings](https://servicebinding.io/application-developer/)).
Moreover users have no control over the annotations that are applied on the function runtime pod. This excludes functions pods from features enabled by annotations (i.e custom log parsers via `fluentbit.io/parser: my-regex-parser`).

## Motivation

Give Serverless users the ability to:
- Configure volume mounted secrets (and config maps) for Function's subresources.
- Configure labels and annotations for the  Function runtime pod.

### Goals

- Add more flexibility to the Serverless API - enable volume mounted secrets and pod annotations.
- Organise spec attributes belonging to runtime and build-time configuration (?)

> - Provide a rollout plan to implement this functionality in small increments and avoid the need to roll out a new API version

Reviewer 4: This is very important, yet not included.

### Non-Goals

> - Define a new API version specification

Reviewer 4: I don't think this would be possible. All the proposals have big changes to the current API, not just additions. I think if we are going to keep this non-goal in mind, we should include a high-level implementation plan on how to achieve that.

## Discussion points

> 

Reviewer 4: Should we look into adding metadata to the build job pod?

Author: included in the last "compromise" option

### Runtime, build-time separation

Since Function CR is managing two workloads ( Deployment for runtime and Job for build-time ) we need to separate them in the spec in order to make it clear where do the mounts and annotations belong. 

We can either make a clear separation, i.e:

```yaml 
spec:
  source:
  runtimeSpec:
    # mounts
    # resources
    # envs
    # metadata
  buildSpec:
    # resources
    # envs (?)
    # metadata (?)
```

Alternatively we could promote runtime fields to the root (as the belong to kind: Function) and extract only the build-time fields 

```yaml 
spec:
  source:
  # mounts
  # resources
  # envs
  # metadata
  buildSpec:
    # resources
    # envs (?)
    # metadata (?)
```

> ### Mounts - own structure or k8s inherited

Reviewer 4: I prefer providing the user with more flexibility and exposing the full volume API, but if we can't, the serviceBinding option sound good.

Author: I discussed it with @pbochynski 

we could separate those cases.
We definitely need a convenient way to enable service bindings.
Those use natively  k8s secrets ( i.e created via BTP service binding CRs ) and leverage dedicated SDKs compliant with  service binding recommendation

volume mounts is more generic but would be less convinient for devs focused on the main use case ( service bindings )

Volume mounts could be a separate feature in our spec ( see last "compromise" option )

Under the hood the secret mount becomes a volume mount in the runtime pod.
We could :
 - expose the k8s volume mount spec  in function spec

>  - cover k8s volume mount spec via oor own [specialised](https://servicebinding.io/application-developer/) spec

Reviewer 5: ```suggestion
 - cover k8s volume mount spec via our own [specialized](https://servicebinding.io/application-developer/) spec
```

>  - cover k8s volume mount spec via oor own [specialised](https://servicebinding.io/application-developer/) spec

Reviewer 5: Do you mean to reuse service binding from BTP service operator or create something new? First will be too limiting, and second too confusing.

Author: I imagine a simple way to define a reference to a secret and optional mounting point (`SERVICE_BINDING_ROOT` env could be used by default ) to support volume mounted secrets as recommended [here](https://servicebinding.io/)
Something like 
```yaml
spec:
  serviceBindings:
  - source: my-redis-secret
  - mountPath: "/redis-secret" # optional
```

Author: discussed. and the outcome reflected in the PR

    secret:
      secretName: mysecret
```

```yaml 
apiVersion: serverless.kyma-project.io/v1alpha2
kind: Function
spec:
  serviceBindings:
  - source: my-secret
  env:
  - name: SERVICE_BINDING_ROOT 
    value: /foo
```

### Metadata for function pod

Define runtime labels and annotations on the root level or under a `metadata` field

```yaml
apiVersion: serverless.kyma-project.io/v1alpha2
kind: Function
metadata:
  name: my-function
  namespace: default
  labels:
    ...
  annotations:
    ...
spec:
  metadata: 
    labels: ...
    annotations: ...
```
OR
```yaml
apiVersion: serverless.kyma-project.io/v1alpha2
kind: Function
metadata:
  name: my-function
  namespace: default
  labels:
    ...
  annotations:
    ...
spec:
  labels: ...
  annotations: ...
```

## Samples: 

### Option 1

- simplified mounts serving just for service binding purpose
- extract `build` object for any build-time specific config
- runtime labels and annotations on the spec root level

```yaml
apiVersion: serverless.kyma-project.io/v1alpha2
kind: Function
metadata:
  name: my-function
  namespace: default
  labels:
    app.kubernetes.io/name: my-function
spec:
  runtime: nodejs16
  source:
    ...

>   serviceBindings: 

Reviewer 4: This is an interesting name. I think it makes sense.

Author: the final name would have to be `secretBindings` though as `service binding` would clash with https://github.com/SAP/sap-btp-service-operator#service-binding

Author: in the end the implicit associations with service bindings (https://github.com/SAP/sap-btp-service-operator#service-binding or https://servicebinding.io/) turn out to be unwanted. 
To keep the API not clean from this domain the proposal was changed to `secretMounts`

  - source: my-secret
    mountPath: "/foo" # optional.. read from SERVICE_BINDING_ROOT ENV 

  profile: S / M / L / XL / ... / Custom #optional
  resources: # optional... required if spec.profile==Custom
    limits: ... #if_custom
    requests: ... #if_custom
​
  
  labels: 
    app: my-app
  annotations: 
    fluentbit.io/parser: my-regex-parser
    istio-injection: enabled
​
  env:
  - name: SERVICE_BINDING_ROOT
    value: /service_bindings
​
​
  build:
    profile: S / M / L / XL / ... / Custom
    resources: # optional... required if spec.build.profile==Custom
      limits: ... #if_custom
      requests: ... #if_custom
​
```
### Option 2

- almost the same as `Option 1` (same pros)

> - dif1: move some fields from `.spec` to the new struct `.spec.runtimeSpec` to clearly distinguish fields desired to be used in the `build` and `running` phases. For example in `Option 1` users may have questions after seeing `.spec.env` and `.spec.build.env` fields for example "is .spec.env dedicated for the running function's pod? Would the field be merged with .spec.build.env for the building job?"

Reviewer 4: Also agree with this. The split is confusing especially considering our current API.

Author: included in the last "compromise option"

      volumes:
        - name: search-index
            nfs:
              path: /path-to-index
              readOnly: true
              server: localhost
        - name: config
          configmap:
            name: function-configuration
            items:
              - key: config
                path: config.yaml
    buildPod:
      metadata: ...# labels and annotations only for build pod
      spec: 
        resources: # optional... required if spec.resourceProfiles.build is empty
          limits: ... #if_custom
          requests: ... #if_custom
        env: ...
        volumeMounts: ...
      volumes: ...
```

### Compromise

 - clearly separate configuration of the build stage in the spec.
 - add simple, [purpose focused](https://blogs.sap.com/2022/07/12/the-new-way-to-consume-service-bindings-on-kyma-runtime/) `serviceBindings` 
 - volume mounts separated to a different place, for more advanced case (implemented once requested) 
 - don't group labels and annotations under metadata. 

```yaml
apiVersion: serverless.kyma-project.io/v1alpha2
kind: Function
metadata:
  name: my-function
  namespace: default
  labels:
    app.kubernetes.io/name: my-function

> spec:

Reviewer 2: ```suggestion
spec: #Contains spec of Function and run stage (deployments, hpa)
```

  runtime: nodejs16
  source:
    ...
  replicas: 1
  scalingConfig:
    min: 1
    max: 2

  labels: #?
    app: my-app
  annotations: #?
    fluentbit.io/parser: my-regex-parser
    istio-injection: enabled

>   serviceBindings: 

Reviewer 6: I'm wondering if that explicit relation to the ServiceBinding concept is good. That will introduce a dependency to it and you would need to explicitly test that (have an app using the specific libraries to interpret the metadata of a binding, to assure that actual serviceBindings can be consumed)
In contrast, if someone just wants to consume a 3party API using secret mounts he will be confused and wonder if that mechanism can be used for that.
So what are the actual parts which are specific to ServiceBindings? As far as I can see the defaulting of the mountPath to the SERVICE_BINDING_ROOT only, right?
We could either drop it fully and except a little loss in convenience

```yaml
secretMounts
  - secretName: icke
    mountPath: "myMountPath"
```
or we could add it as a capability to the binding itself?

```yaml
secretMounts
  - secretName: icke
     isServiceBinding: true
```

or more generic but coupled:
```yaml
secretMounts
  - fromSecret:
       name: icke
     mountPath: "/icke"
  - fromServiceBinding:
       name: "icke"
```

Author: Thank you @a-thaler 
In order to avoid implicit association with service bindings (https://github.com/SAP/sap-btp-service-operator#service-binding or https://servicebinding.io/) and user confusion I think the best option would be to simply enable the secret mount.

If user wants to use specialised libraries we would need to extra set the env variable, for example:

```yaml
secretMounts
  - secretName: icke
    mountPath: "/foo" #required.. no assumptions/validations towards SERVICE_BINDING_ROOT env value
    
 envs:
  - name: SERVICE_BINDING_ROOT
     value: "/foo"  #set explicitely if user wants to use a specialised library that expects it      
```

Author: the fact that `mountPath` and SERVICE_BINDING_ROOT value match indicates that this secret mount will be consumed in an "Service binding"- way via specialised library. 
But this is soft indication, not done by our API.
I think its sufficiently convenient

  - secretName: my-secret

>     mountPath: "/foo" # optional.. read from SERVICE_BINDING_ROOT ENV 

Reviewer 2: ```suggestion
    mountPath: "/foo" # optional.. if not set read from SERVICE_BINDING_ROOT ENV 
```

Reviewer 6: And SERVICE_BINDING_ROOT env is set by default?

Author: `mountPath` is required.
`SERVICE_BINDING_ROOT` ENV  is optional and used by users who are willing to use specialised libraries to consume mounted secrets as service bindings

>   profile: S / M / L / XL / ... / Custom #optional

Reviewer 2: I wouldn't use `Custom` to enable manual resoruces. The simplier to use option is to leave it empty.

Reviewer 6: and fail validation if both are specified

Author: included in the PR

  resources: # optional... required if spec.profile==Custom
    limits: ... #if_custom
    requests: ... #if_custom
​
  env:
  - name: SERVICE_BINDING_ROOT
    value: /service_bindings
  - name: MODE
    value: modeA

  # volumeMounts:  (add when requested by users)
  # - name: config
  #   mountPath: /etc/config/
  # - name: search-index
  #   mountPath: /etc/index

  # volumes:
  # - name: search-index
  #     nfs:
  #       path: /path-to-index
  #       readOnly: true
  #       server: localhost
  # - name: config
  #   configmap:
  #     name: function-configuration
  #     items:
  #       - key: config
  #         path: config.yaml
​​

  # optional 

>   build:

Reviewer 2: ```suggestion
  build: #Contains spec of build stage
```

            items:
              - key: config
                path: config.yaml
    buildPod:
      metadata: ...# labels and annotations only for build pod
      spec: 
        resources: # optional... required if spec.resourceProfiles.build is empty
          limits: ... #if_custom
          requests: ... #if_custom
        env: ...
        volumeMounts: ...
      volumes: ...
```

### Final version - the compromise

 - clearly separate configuration of the build stage in the spec (treating `build` stage as second class citizen and keeping the main spec dedicated to the more important runtime stage).
 - add convinient way to mount secrets (w/o polluting function API with dependencies to service bindings)
 - volume mounts as a separate, more advanced case (implemented once requested) 
 - don't group labels and annotations under metadata. 

```yaml
apiVersion: serverless.kyma-project.io/v1alpha2
kind: Function
metadata:
  name: my-function
  namespace: default
  labels:
    app.kubernetes.io/name: my-function
spec: #Contains spec of Function and run stage (deployments, hpa)
  runtime: nodejs16
  source:
    ...
  replicas: 1
  scalingConfig:
    min: 1
    max: 2

>   labels: #?

Reviewer 2: ```suggestion
  labels:
```

    app: my-app

>   annotations: #?

Reviewer 2: ```suggestion
  annotations:
```

    requests: ... #if profile empty
​
  env:
  - name: SERVICE_BINDING_ROOT #set explicitely by user if he wants to use a specialised library that expects the ENV (and allows consumption of mounted secrets as service bindings)
    value: /bar
  - name: MODE
    value: modeA

  # volumeMounts:  (add when requested by users)
  # - name: config
  #   mountPath: /etc/config/
  # - name: search-index
  #   mountPath: /etc/index

  # volumes:
  # - name: search-index
  #     nfs:
  #       path: /path-to-index
  #       readOnly: true
  #       server: localhost
  # - name: config
  #   configmap:
  #     name: function-configuration
  #     items:
  #       - key: config
  #         path: config.yaml
​​

  # optional 
  build: #Contains spec of build stage : build is a "second class citizen" here. Users should make no assumptions that anything from main spec is inherited here (i.e ENVs or secretMounts)
    labels: 
    annotations: 
    profile: S / M / L / XL / ... #optional
    resources: # optional... required if spec.profile is empty
      limits: ... #if profile empty
      requests: ... #if profile empty

```

### Precedence, defaulting and validation

> `Profile` takes precedence over `resources`. If profile field is not set to "Custom" there should be no defaulting happening for the resources. Controller should fill the pod template resources accoriding to the selected profile preset.

Reviewer 2: ```suggestion
`profile` takes precedence over `resources`. If profile field is not set there should be no defaulting happening for the resources. Controller should fill the pod template resources according to the selected profile preset.
```

# Extending CRD spec for Serverless v1alpha2

## Summary

The current Serverless API allows for limited configuration of the generated Function's deployment. 

> Currently users can only use ENVs to pass 3rd party service credentials but it doesn't allow for volume mounted secrets (which becomes the industry standard for [service bindings](https://servicebinding.io/application-developer/)).

Reviewer 7: ```suggestion
Currently, users can only use ENVs to pass 3rd party service credentials but it doesn't allow for volume-mounted Secrets (which become the industry standard for [service bindings](https://servicebinding.io/application-developer/)).
```

> Moreover users have no control over the annotations that are applied on the function runtime pod. This excludes functions pods from features enabled by annotations (i.e custom log parsers via `fluentbit.io/parser: my-regex-parser`).

Reviewer 7: ```suggestion
Moreover, users have no control over the annotations applied on the function runtime pod. This excludes Function's Pods from features enabled by annotations (i.e., custom log parsers via `fluentbit.io/parser: my-regex-parser`).
```

## Motivation

Give Serverless users the ability to:

> - Configure volume mounted secrets for Function's subresources.

Reviewer 7: ```suggestion
- Configure volume-mounted Secrets for Function's subresources.
```

> - Configure labels and annotations for the Function runtime pod.

Reviewer 7: ```suggestion
- Configure labels and annotations for the Function's runtime Pod.
```

### Goals

> - Add more flexibility to the Serverless API - enable volume mounted secrets and pod annotations.

Reviewer 7: ```suggestion
- Add more flexibility to the Serverless API - enable volume-mounted Secrets and Pod annotations.
```

- Organise spec attributes belonging to runtime and build-time configuration (?)
- Propose sample Function CR visualising different variants

## Discussion points

### Runtime, build-time separation

> Since Function CR is managing two workloads ( Deployment for runtime and Job for build-time ) we need to separate them in the spec in order to make it clear where do the mounts and annotations belong. 

Reviewer 7: ```suggestion
Since Function CR is managing two workloads (Deployment for runtime and Job for build-time) we need to separate them in the spec in order to make it clear where the mounts and annotations belong. 
```

> We can either make a clear separation, i.e:

Reviewer 7: ```suggestion
We can either make a clear separation, i.e.:
```

```yaml 
spec:
  source:
  runtimeSpec:
    # mounts
    # resources
    # envs
    # metadata
  buildSpec:
    # resources
    # envs (?)
    # metadata (?)
```

> Alternatively we could promote runtime fields to the root (as the belong to kind: Function) and extract only the build-time fields 

Reviewer 7: ```suggestion
Alternatively, we could promote runtime fields to the root (as they belong to kind: Function) and extract only the build-time fields 
```

```yaml 
spec:
  source:
  # mounts
  # resources
  # envs
  # metadata
  buildSpec:
    # resources
    # envs (?)
    # metadata (?)
```

### Mounts - own structure or k8s inherited

> Under the hood the secret mount becomes a volume mount in the runtime pod.

Reviewer 7: ```suggestion
Under the hood, the Secret mount becomes a volume mount in the runtime Pod.
```

> We could :

Reviewer 7: ```suggestion
We could:
```

> A) expose the k8s volume mount spec in function spec

Reviewer 7: ```suggestion
A) expose the k8s volume mount spec in the Function spec
```

Pros: 

>  - Generic solution - Allows mounting anything : secrets, config maps, any volumes

Reviewer 7: ```suggestion
 - Generic solution - Allows mounting anything: Secrets, ConfigMaps, any volumes
```

>  - very easy to achieve (rewriting from function spec to pod teplate spec)

Reviewer 7: ```suggestion
 - very easy to achieve (rewriting from the Function spec to Pod template spec)
```

Cons: 

>  - Does not represent the service binging use case. User needs to translate service bindngs into volume mounts by themselves

Reviewer 7: ```suggestion
 - Does not represent the service binding use case. User needs to translate service bindings into volume mounts by themselves
```

>  - Less compact ( elegant ) to configure requested service binding use case

Reviewer 7: ```suggestion
 - Less compact (elegant) to configure requested service binding use case
```

>  - Noone requested it yet

Reviewer 7: ```suggestion
 - Noone has requested it yet
```

```yaml 
apiVersion: serverless.kyma-project.io/v1alpha2
kind: Function
spec:
  volumeMounts:
  - name: foo
    mountPath: "/etc/foo"
    readOnly: true
  volumes:
  - name: foo
    secret:
      secretName: mysecret
```
 B) focus on [service binding case](https://servicebinding.io/application-developer/).

```yaml 
apiVersion: serverless.kyma-project.io/v1alpha2
kind: Function
spec:
  secretBindings: #serviceBindings will cause name clash with https://github.com/SAP/sap-btp-service-operator#service-binding
  - source: my-secret
    mountPath: /bar # optional mount path
  env:
  - name: SERVICE_BINDING_ROOT # default mount path for service bindings
    value: /foo
```
 Pros: 
 - Purpose focused  
 - Compact configuration - easy to adopt

>  - Less confusing (as volume mounts cause confusion as serverless functions are considered stateless and should not claim any persistance volumes )

Reviewer 7: ```suggestion
 - Less confusing (as volume mounts confuse Serverless Functions are considered stateless and should not claim any persistence volumes )
```

 - Enables using service binding natively with dedicated SDKs (i.e @sap/xsenv) - [related read](https://blogs.sap.com/2022/07/12/the-new-way-to-consume-service-bindings-on-kyma-runtime/)

Cons: 

>  - Not allows to mount anything besides secrets

Reviewer 7: ```suggestion
 - Not allows mounting anything besides Secrets
```

A) and B) are not exclusive

We could separate those cases. (See last 'compromise' option)

> ### Metadata for function pod

Reviewer 7: ```suggestion
### Metadata for Function's Pod
```

- stretch: volume mounts could be added as a separate feature
- extract `build` object for any build-time specific config
- runtime labels and annotations on the spec root level

```yaml
apiVersion: serverless.kyma-project.io/v1alpha2
kind: Function
metadata:
  name: my-function
  namespace: default
  labels:
    app.kubernetes.io/name: my-function
spec:
  runtime: nodejs16
  source:
    ...
  secretBindings: 
  - source: my-secret
    mountPath: "/foo" # optional.. read from SERVICE_BINDING_ROOT ENV 

  profile: S / M / L / XL / ... / Custom #optional
  resources: # optional... required if spec.profile==Custom
    limits: ... #if_custom
    requests: ... #if_custom
​  
  labels: 
    app: my-app
  annotations: 
    fluentbit.io/parser: my-regex-parser
    istio-injection: enabled
​
  env:
  - name: SERVICE_BINDING_ROOT
    value: /service_bindings
​​
  build:
    profile: S / M / L / XL / ... / Custom
    resources: # optional... required if spec.build.profile==Custom
      limits: ... #if_custom
      requests: ... #if_custom
    labels: #optional
    annotations: #optional
```

### Option 2

- almost the same as `Option 1` (same pros)

> - dif1: move some fields from `.spec` to the new struct `.spec.runtimeSpec` to clearly distinguish fields desired to be used in the `build` and `running` phases. For example in `Option 1` users may have questions after seeing `.spec.env` and `.spec.build.env` fields for example "is .spec.env dedicated for the running function's pod? Would the field be merged with .spec.build.env for the building job?"

Reviewer 7: ```suggestion
- dif1: move some fields from `.spec` to the new struct `.spec.runtimeSpec` to clearly distinguish fields desired to be used in the `build` and `running` phases. For example in `Option 1` users may have questions after seeing `.spec.env` and `.spec.build.env` fields for example "is .spec.env dedicated for the running Function's Pod? Would the field be merged with .spec.build.env for the building job?"
```

- dif2: rename the `.spec.profile` to the `.spec.resourcesProfile` to make this field more intuitive
- dif3: this solution is simple but not as simple as `Option 1`

> >NOTE: the main idea is to close a specific configuration in a field that represents the specific phase of the function's lifecycle. It would be intuitive and easy to understand for a user that the `.spec.build` field contains configuration for the building phase and `.spec.runtimeSpec` for the running phase.

Reviewer 7: ```suggestion
>NOTE: the main idea is to close a specific configuration in a field that represents the specific phase of the Function's lifecycle. It would be intuitive and easy to understand for a user that the `.spec.build` field contains configuration for the building phase and `.spec.runtimeSpec` for the running phase.
```

```yaml
apiVersion: serverless.kyma-project.io/v1alpha2
kind: Function
metadata:
  name: my-function
  namespace: default
  labels:
    app.kubernetes.io/name: my-function
spec:
  runtime: nodejs16
  source:
    ...
  runtimeSpec: # this name is not perfect
    serviceBindings: 
    - source: my-secret
      mountPath: "/foo" # optional.. read from SERVICE_BINDING_ROOT ENV 

    resourcesProfile: S / M / L / XL / ... / Custom #optional
    resources: # optional... required if spec.profile==Custom
      limits: ... #if_custom
      requests: ... #if_custom

    labels: 
      app: my-app
    annotations: 
      fluentbit.io/parser: my-regex-parser
      istio-injection: enabled
  
    env:
    - name: SERVICE_BINDING_ROOT
      value: /service_bindings

  buildSpec: # this name is not perfect
    resourcesProfile: S / M / L / XL / ... / Custom
    resources: # optional... required if spec.build.profile==Custom
      limits: ... #if_custom
      requests: ... #if_custom

```

### Option 3

> This option expose runtime pod configuration over build pod because runtime pod is final result, the build is transient phase.

Reviewer 7: ```suggestion
This option exposes runtime pod configuration over the build Pod because the runtime Pod is the final result, and the build is a transient phase.
```

                key: password
          - name: EXTERNAL_API_URL
            valueFrom:
              configmapRef:
                name: mysvc-configuration
                key: URL
        volumeMounts: ...
          - name: config
            mountPath: /etc/config/
          - name: search-index
            mountPath: /etc/index
      volumes:
        - name: search-index
            nfs:
              path: /path-to-index
              readOnly: true
              server: localhost
        - name: config
          configmap:
            name: function-configuration
            items:
              - key: config
                path: config.yaml
    buildPod:
      metadata: ...# labels and annotations only for build pod
      spec: 
        resources: # optional... required if spec.resourceProfiles.build is empty
          limits: ... #if_custom
          requests: ... #if_custom
        env: ...
        volumeMounts: ...
      volumes: ...
```

### Final version - the compromise

 - clearly separate configuration of the build stage in the spec (treating `build` stage as second class citizen and keeping the main spec dedicated to the more important runtime stage).

>  - add convinient way to mount secrets (w/o polluting function API with dependencies to service bindings)

Reviewer 7: ```suggestion
 - add convenient way to mount Secrets (w/o polluting function API with dependencies to service bindings)
```

    requests: ... #if profile empty
​
  env:
  - name: SERVICE_BINDING_ROOT #set explicitely by user if he wants to use a specialised library that expects the ENV (and allows consumption of mounted secrets as service bindings)
    value: /bar
  - name: MODE
    value: modeA

  # volumeMounts:  (add when requested by users)
  # - name: config
  #   mountPath: /etc/config/
  # - name: search-index
  #   mountPath: /etc/index

  # volumes:
  # - name: search-index
  #     nfs:
  #       path: /path-to-index
  #       readOnly: true
  #       server: localhost
  # - name: config
  #   configmap:
  #     name: function-configuration
  #     items:
  #       - key: config
  #         path: config.yaml
​​

  # optional 
  build: #Contains spec of build stage : build is a "second class citizen" here. Users should make no assumptions that anything from main spec is inherited here (i.e ENVs or secretMounts)
    labels: 
    annotations: 
    profile: S / M / L / XL / ... #optional
    resources: # optional... required if spec.profile is empty
      limits: ... #if profile empty
      requests: ... #if profile empty

```

### Precedence, defaulting and validation

> `profile` takes precedence over `resources`. If profile field is not set there should be no defaulting happening for the resources. Controller should fill the pod template resources according to the selected profile preset.

Reviewer 7: ```suggestion
`profile` takes precedence over `resources`. If the profile field is not set there should be no defaulting happening for the resources. The controller should fill the pod template resources according to the selected profile preset.
```

> If profile field is set to "Custom", the user must then ( and only then ) set the values for resources manualy.

Reviewer 7: ```suggestion
If the profile field is set to "Custom", the user must then (and only then) set the values for resources manually.
```
